src/mainframe/job_cache.py

```diff
@@ -0,0 +1,224 @@
+                    "Timed out package found. Requeueing.", name=pending_scan.name, version=pending_scan.version
+                )
+
+        return scans
+
+    def refill(self) -> None:
+        # refill from timed out pending scans first
+        logger.info("Refilling from timed out pending scans")
+        requeued_scans = self.requeue_timeouts()
+        logger.info(f"Moved {len(requeued_scans)} timed out scans from pending to queue")
+
+        # exclude packages that we just put back into the queue
+        # and packages that are currently pending from our database query
+        requeued_scans_name_version = [(s.name, s.version) for s in requeued_scans]
+        pending_name_version = [(s.name, s.version) for s in self.pending]
+        scans_to_exclude = requeued_scans_name_version + pending_name_version
+
+        query = (
+            select(Scan)
+            .where(Scan.status == Status.QUEUED)
+            .where(tuple_(Scan.name, Scan.version).not_in(scans_to_exclude))
+            .order_by(Scan.queued_at)
+            .limit(self.scan_queue.maxsize)
+            .options(selectinload(Scan.rules), selectinload(Scan.download_urls))
+        )
+
+        scans = self.session.scalars(query).all()
+
+        logger.info(f"Fetched {len(scans)} scans from DB to refill queue with.")
+
+        for scan in scans:
+            try:
+                self.scan_queue.put(scan, timeout=5)
+                logger.info("Put scan into queue.", name=scan.name, version=scan.version)
+            except queue.Full:
+                # this scenario can happen if some jobs from the timeout have already
+                # been added into the queue. In this case we just ignore the remaining
+                # jobs from the DB since we're already full
+                logger.warn("Overfetched. Ignoring extras.")
+                break
+
+        logger.info("Refilled jobs queue")
+
+    def persist_all_results(self) -> None:
+        """Pop off all results and persist them in the database"""
+        results: list[PackageScanResult | PackageScanResultFail] = []
+        while not self.results_queue.empty():
+            result = self.results_queue.get(timeout=5)
+            results.append(result)
+
+        name_versions = [(result.name, result.version) for result in results]
+        query = select(Scan).where(tuple_(Scan.name, Scan.version).in_(name_versions))
+        scans = self.session.scalars(query).all()
+
+        all_rules = self.session.scalars(select(Rule)).all()
+
+        for result in results:
+            scan = next((scan for scan in scans if (scan.name, scan.version) == (result.name, result.version)), None)
+            if scan is None:
+                logger.warn("Results submitted for a package that doesn't exist, skipping", **result.model_dump())
+                continue
+
+            if scan.status == Status.FINISHED:
+                logger.warn("Package is already in a FINISHED state, skipping", **result.model_dump())
```

Comment:
I would add a `continue` here to make it clear you're skipping the rest of the loop.